File: src/Historical/Pascal/singleton.pas
program SingletonExample;

{$mode objfpc}{$H+}

type
  PRegistry = ^TRegistry;
  TRegistry = record
    Count: Integer;
  end;

var
  SharedRegistry: TRegistry;

function Instance: PRegistry;
begin
  Result := @SharedRegistry;
end;

procedure Increment(Registry: PRegistry);
begin
  Inc(Registry^.Count);
end;

var
  FirstRegistry: PRegistry;
  SecondRegistry: PRegistry;
begin
  FirstRegistry := Instance;
  SecondRegistry := Instance;
  Increment(FirstRegistry);

  if FirstRegistry = SecondRegistry then
    WriteLn('same=true')
  else
    WriteLn('same=false');
  WriteLn('count=', SecondRegistry^.Count);
end.

File: src/Historical/Fortran/singleton.f90
module registry_module
  implicit none
  type :: registry_state
    integer :: count = 0
  end type registry_state
  type(registry_state), target, save :: shared
contains
  function instance() result(value)
    type(registry_state), pointer :: value
    value => shared
  end function instance
end module registry_module

program singleton
  use registry_module
  implicit none
  type(registry_state), pointer :: first, second

  first => instance()
  second => instance()
  first%count = first%count + 1

  if (associated(first, second)) then
    print '(A)', 'same=true'
  else
    print '(A)', 'same=false'
  end if
  print '(A,I0)', 'count=', second%count
end program singleton
